File: src/main/java/net/aspw/client/util/newfont/CFontRenderer.java
package net.aspw.client.util.newfont;

import java.awt.Color;
import java.awt.Font;
import java.util.ArrayList;
import java.util.List;

import net.aspw.client.util.render.ColorUtils;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.texture.DynamicTexture;
import org.lwjgl.opengl.GL11;

public class CFontRenderer extends CFont {
    protected CFont.CharData[] boldChars = new CFont.CharData[256];
    protected CFont.CharData[] italicChars = new CFont.CharData[256];
    protected CFont.CharData[] boldItalicChars = new CFont.CharData[256];
    private final int[] colorCode = new int[32];
    private final String colorcodeIdentifiers = "0123456789abcdefklmnor";
    protected DynamicTexture texBold;
    protected DynamicTexture texItalic;
    protected DynamicTexture texItalicBold;

    public CFontRenderer(Font font, boolean antiAlias, boolean fractionalMetrics) {
        super(font, antiAlias, fractionalMetrics);
        this.setupMinecraftColorcodes();
        this.setupBoldItalicIDs();
    }

    public float drawStringWithShadow(String text, double x, double y, int color) {
        float shadowWidth = drawString(text, x + 0.5D, y + 0.5D, color, true);
        return Math.max(shadowWidth, drawString(text, x, y, color, false));
    }

    public void drawStringWithOutline(String str, float x, float y, int color, int outline) {
        this.drawString(ColorUtils.stripColor(str), x - 1f, y, outline, false);
        this.drawString(ColorUtils.stripColor(str), x + 1f, y, outline, false);
        this.drawString(ColorUtils.stripColor(str), x, y + 1f, outline, false);
        this.drawString(ColorUtils.stripColor(str), x, y - 1f, outline, false);
        this.drawString(ColorUtils.stripColor(str), x, y, color, false);
    }

    public float drawString(String text, float x, float y, int color) {
        return this.drawString(text, x, y, color, false);
    }

    public float drawCenteredString(String text, float x, float y, int color) {
        return this.drawString(text, x - (float) (this.getStringWidth(text) / 2) - 1, y, color);
    }

    public float drawCenteredStringWithShadow(String text, float x, float y, int color) {
        return this.drawStringWithShadow(text, x - (float) (this.getStringWidth(text) / 2) - 1, y, color);
    }

    public float drawCenteredStringWithShadow(String text, double x, double y, int color) {
        return this.drawStringWithShadow(text, x - (double) (this.getStringWidth(text) / 2) - 1, y, color);
    }

    public float drawString(final String text, double x, double y, int color, final boolean shadow) {
        --x;
        if (text == null) {
            return 0.0f;
        }
        if (color == 553648127) {
            color = 16777215;
        }
        if ((color & 0xFC000000) == 0x0) {
            color |= 0xFF000000;
        }
        if (shadow) {
            color = ((color & 0xFCFCFC) >> 2 | (color & new Color(20, 20, 20, 200).getRGB()));
        }
        CharData[] currentData = this.charData;
        final float alpha = (color >> 24 & 0xFF) / 255.0f;
        boolean bold = false;
        boolean italic = false;
        boolean strikethrough = false;
        boolean underline = false;
        x *= 2.0;
        y = (y - 3.0) * 2.0;
        GL11.glPushMatrix();
        GlStateManager.scale(0.5, 0.5, 0.5);
        GlStateManager.enableBlend();
        GlStateManager.blendFunc(770, 771);
        GlStateManager.color((color >> 16 & 0xFF) / 255.0f, (color >> 8 & 0xFF) / 255.0f, (color & 0xFF) / 255.0f,
                alpha);
        final int size = text.length();
        GlStateManager.enableTexture2D();
        GlStateManager.bindTexture(this.tex.getGlTextureId());
        GL11.glBindTexture(3553, this.tex.getGlTextureId());
        for (int i = 0; i < size; ++i) {
            final char character = text.charAt(i);
            if (String.valueOf(character).equals("\u00a7")) {
                int colorIndex = 21;
                try {
                    colorIndex = "0123456789abcdefklmnor".indexOf(text.charAt(i + 1));
                } catch (Exception e) {
                    e.printStackTrace();
                }
                if (colorIndex < 16) {
                    bold = false;
                    italic = false;
                    underline = false;
                    strikethrough = false;
                    GlStateManager.bindTexture(this.tex.getGlTextureId());
                    currentData = this.charData;
                    if (colorIndex < 0) {
                        colorIndex = 15;
                    }
                    if (shadow) {
                        colorIndex += 16;
                    }
                    final int colorcode = this.colorCode[colorIndex];
                    GlStateManager.color((colorcode >> 16 & 0xFF) / 255.0f, (colorcode >> 8 & 0xFF) / 255.0f,
                            (colorcode & 0xFF) / 255.0f, alpha);
                } else if (colorIndex == 17) {
                    bold = true;
                    if (italic) {
                        GlStateManager.bindTexture(this.texItalicBold.getGlTextureId());
                        currentData = this.boldItalicChars;
                    } else {
                        GlStateManager.bindTexture(this.texBold.getGlTextureId());
                        currentData = this.boldChars;
                    }
                } else if (colorIndex == 18) {
                    strikethrough = true;
                } else if (colorIndex == 19) {
                    underline = true;
                } else if (colorIndex == 20) {
                    italic = true;
                    if (bold) {
                        GlStateManager.bindTexture(this.texItalicBold.getGlTextureId());
                        currentData = this.boldItalicChars;
                    } else {
                        GlStateManager.bindTexture(this.texItalic.getGlTextureId());
                        currentData = this.italicChars;
                    }
                } else if (colorIndex == 21) {
                    bold = false;
                    italic = false;
                    underline = false;
                    strikethrough = false;
                    GlStateManager.color((color >> 16 & 0xFF) / 255.0f, (color >> 8 & 0xFF) / 255.0f,
                            (color & 0xFF) / 255.0f, alpha);
                    GlStateManager.bindTexture(this.tex.getGlTextureId());
                    currentData = this.charData;
                }
                ++i;
            } else if (character < currentData.length) {
                GL11.glBegin(4);
                this.drawChar(currentData, character, (float) x, (float) y);
                GL11.glEnd();
                if (strikethrough) {
                    this.drawLine(x, y + currentData[character].height / 2.0f, x + currentData[character].width - 8.0,
                            y + currentData[character].height / 2.0f, 1.0f);
                }
                if (underline) {
                    this.drawLine(x, y + currentData[character].height - 2.0, x + currentData[character].width - 8.0,
                            y + currentData[character].height - 2.0, 1.0f);
                }
                x += currentData[character].width - 9 + this.charOffset;
            }
        }
        GL11.glPopMatrix();
        return (float) x / 2.0f;
    }

    @Override
    public int getStringWidth(String text) {
        if (text == null) {
            return 0;
        }
        int width = 0;
        CFont.CharData[] currentData = this.charData;
        boolean bold = false;
        boolean italic = false;
        int size = text.length();
        int i = 0;
        while (i < size) {
            char character = text.charAt(i);
            if (character == '\u00a7' && i < size) {
                int colorIndex = "0123456789abcdefklmnor".indexOf(character);
                if (colorIndex < 16) {
                    bold = false;
                    italic = false;
                } else if (colorIndex == 17) {
                    bold = true;
                    currentData = italic ? this.boldItalicChars : this.boldChars;
                } else if (colorIndex == 20) {
                    italic = true;
                    currentData = bold ? this.boldItalicChars : this.italicChars;
                } else if (colorIndex == 21) {
                    bold = false;
                    italic = false;
                    currentData = this.charData;
                }
                ++i;
            } else if (character < currentData.length && character >= '\u0000') {
                width += currentData[character].width - 9 + this.charOffset;
            }
            ++i;
        }
        return width / 2;
    }

    @Override
    public void setFont(Font font) {
        super.setFont(font);
        this.setupBoldItalicIDs();
    }

    @Override
    public void setAntiAlias(boolean antiAlias) {
        super.setAntiAlias(antiAlias);
        this.setupBoldItalicIDs();
    }

    @Override
    public void setFractionalMetrics(boolean fractionalMetrics) {
        super.setFractionalMetrics(fractionalMetrics);
        this.setupBoldItalicIDs();
    }

    private void setupBoldItalicIDs() {
        this.texBold = this.setupTexture(this.font.deriveFont(1), this.antiAlias, this.fractionalMetrics,
                this.boldChars);
        this.texItalic = this.setupTexture(this.font.deriveFont(2), this.antiAlias, this.fractionalMetrics,
                this.italicChars);
    }

    private void drawLine(double x, double y, double x1, double y1, float width) {
        GL11.glDisable((int) 3553);
        GL11.glLineWidth((float) width);
        GL11.glBegin((int) 1);
        GL11.glVertex2d((double) x, (double) y);
        GL11.glVertex2d((double) x1, (double) y1);
        GL11.glEnd();
        GL11.glEnable((int) 3553);
    }

    public List<String> wrapWords(String text, double width) {
        ArrayList<String> finalWords = new ArrayList<String>();
        if ((double) this.getStringWidth(text) > width) {
            String[] words = text.split(" ");
            String currentWord = "";
            int lastColorCode = 65535;
            String[] arrstring = words;
            int n = arrstring.length;
            int n2 = 0;
            while (n2 < n) {
                String word = arrstring[n2];
                int i = 0;
                while (i < word.toCharArray().length) {
                    char c = word.toCharArray()[i];
                    if (c == '\u00a7' && i < word.toCharArray().length - 1) {
                        lastColorCode = word.toCharArray()[i + 1];
                    }
                    ++i;
                }
                if ((double) this.getStringWidth(String.valueOf(currentWord) + word + " ") < width) {
                    currentWord = String.valueOf(currentWord) + word + " ";
                } else {
                    finalWords.add(currentWord);
                    currentWord = String.valueOf(167 + lastColorCode) + word + " ";
                }
                ++n2;
            }
            if (currentWord.length() > 0) {
                if ((double) this.getStringWidth(currentWord) < width) {
                    finalWords.add(String.valueOf(167 + lastColorCode) + currentWord + " ");
                    currentWord = "";
                } else {
                    for (String s : this.formatString(currentWord, width)) {
                        finalWords.add(s);
                    }
                }
            }
        } else {
            finalWords.add(text);
        }
        return finalWords;
    }

    public List<String> formatString(String string, double width) {
        ArrayList<String> finalWords = new ArrayList<String>();
        String currentWord = "";
        int lastColorCode = 65535;
        char[] chars = string.toCharArray();
        int i = 0;
        while (i < chars.length) {
            char c = chars[i];
            if (c == '\u00a7' && i < chars.length - 1) {
                lastColorCode = chars[i + 1];
            }
            if ((double) this.getStringWidth(String.valueOf(currentWord) + c) < width) {
                currentWord = String.valueOf(currentWord) + c;
            } else {
                finalWords.add(currentWord);
                currentWord = String.valueOf(167 + lastColorCode) + String.valueOf(c);
            }
            ++i;
        }
        if (currentWord.length() > 0) {
            finalWords.add(currentWord);
        }
        return finalWords;
    }

    private void setupMinecraftColorcodes() {
        int index = 0;
        while (index < 32) {
            int noClue = (index >> 3 & 1) * 85;
            int red = (index >> 2 & 1) * 170 + noClue;
            int green = (index >> 1 & 1) * 170 + noClue;
            int blue = (index >> 0 & 1) * 170 + noClue;
            if (index == 6) {
                red += 85;
            }
            if (index >= 16) {
                red /= 4;
                green /= 4;
                blue /= 4;
            }
            this.colorCode[index] = (red & 255) << 16 | (green & 255) << 8 | blue & 255;
            ++index;
        }
    }

    public String trimStringToWidth(CharSequence text, int width, boolean reverse) {
        StringBuilder builder = new StringBuilder();

        float f = 0.0F;
        int i = reverse ? text.length() - 1 : 0;
        int j = reverse ? -1 : 1;
        boolean flag = false;
        boolean flag1 = false;

        for (int k = i; k >= 0 && k < text.length() && f < width; k += j) {
            char c0 = text.charAt(k);
            float f1 = getStringWidth(String.valueOf(c0));

            if (flag) {
                flag = false;

                if (c0 != 'l' && c0 != 'L') {
                    if (c0 == 'r' || c0 == 'R') {
                        flag1 = false;
                    }
                } else {
                    flag1 = true;
                }
            } else if (f1 < 0.0F) {
                flag = true;
            } else {
                f += f1;
                if (flag1)
                    ++f;
            }

            if (f > width)
                break;

            if (reverse) {
                builder.insert(0, c0);
            } else {
                builder.append(c0);
            }
        }

        return builder.toString();
    }

}

File: src/main/java/net/aspw/client/visual/font/smooth/CFontRenderer.kt
package net.aspw.client.visual.font.smooth

import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.texture.DynamicTexture
import org.lwjgl.opengl.GL11
import java.awt.Color
import java.awt.Font

class CFontRenderer(font: Font?, antiAlias: Boolean, fractionalMetrics: Boolean) :
    CFont(font, antiAlias, fractionalMetrics) {
    protected var boldChars = arrayOfNulls<CharData>(256)
    protected var italicChars = arrayOfNulls<CharData>(256)
    protected var boldItalicChars = arrayOfNulls<CharData>(256)
    private val colorCode = IntArray(32)
    protected var texBold: DynamicTexture? = null
    protected var texItalic: DynamicTexture? = null
    protected var texItalicBold: DynamicTexture? = null

    init {
        setupMinecraftColorcodes()
        setupBoldItalicIDs()
    }

    fun drawString(text: String?, x: Float, y: Float, color: Int): Float {
        return this.drawString(text, x.toDouble(), y.toDouble(), color, false)
    }

    fun drawStringWithShadow(text: String?, x: Double, y: Double, color: Int): Float {
        val shadowWidth = drawString(text, x + 0.5, y + 0.5, color, true)
        return Math.max(shadowWidth, drawString(text, x, y, color, false))
    }

    fun drawCenteredString(text: String, x: Float, y: Float, color: Int): Float {
        return this.drawString(text, x - (getStringWidth(text) / 2).toFloat() - 1, y, color)
    }

    fun drawCenteredStringWithShadow(text: String, x: Float, y: Float, color: Int): Float {
        return drawStringWithShadow(
            text,
            (x - (getStringWidth(text) / 2).toFloat() - 1).toDouble(),
            y.toDouble(),
            color
        )
    }

    fun drawString(text: String?, x: Double, y: Double, color: Int, shadow: Boolean): Float {
        var x = x
        var y = y
        var color = color
        --x
        if (text == null) {
            return 0.0f
        }
        if (color == 553648127) {
            color = 16777215
        }
        if (color and -0x4000000 == 0x0) {
            color = color or -0x1000000
        }
        if (shadow) {
            color = color and 0xFCFCFC shr 2 or (color and Color(20, 20, 20, 200).rgb)
        }
        var currentData = charData
        val alpha = (color shr 24 and 0xFF) / 255.0f
        var bold = false
        var italic = false
        var strikethrough = false
        var underline = false
        x *= 2.0
        y = (y - 3.0) * 2.0
        GL11.glPushMatrix()
        GlStateManager.scale(0.5, 0.5, 0.5)
        GlStateManager.enableBlend()
        GlStateManager.blendFunc(770, 771)
        GlStateManager.color(
            (color shr 16 and 0xFF) / 255.0f, (color shr 8 and 0xFF) / 255.0f, (color and 0xFF) / 255.0f,
            alpha
        )
        val size = text.length
        GlStateManager.enableTexture2D()
        GlStateManager.bindTexture(tex.glTextureId)
        GL11.glBindTexture(3553, tex.glTextureId)
        var i = 0
        while (i < size) {
            val character = text[i]
            if (character.toString() == "\u00a7") {
                var colorIndex = 21
                try {
                    colorIndex = "0123456789abcdefklmnor".indexOf(text[i + 1])
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                if (colorIndex < 16) {
                    bold = false
                    italic = false
                    underline = false
                    strikethrough = false
                    GlStateManager.bindTexture(tex.glTextureId)
                    currentData = charData
                    if (colorIndex < 0) {
                        colorIndex = 15
                    }
                    if (shadow) {
                        colorIndex += 16
                    }
                    val colorcode = colorCode[colorIndex]
                    GlStateManager.color(
                        (colorcode shr 16 and 0xFF) / 255.0f, (colorcode shr 8 and 0xFF) / 255.0f,
                        (colorcode and 0xFF) / 255.0f, alpha
                    )
                } else if (colorIndex == 17) {
                    bold = true
                    currentData = if (italic) {
                        GlStateManager.bindTexture(texItalicBold!!.glTextureId)
                        boldItalicChars
                    } else {
                        GlStateManager.bindTexture(texBold!!.glTextureId)
                        boldChars
                    }
                } else if (colorIndex == 18) {
                    strikethrough = true
                } else if (colorIndex == 19) {
                    underline = true
                } else if (colorIndex == 20) {
                    italic = true
                    currentData = if (bold) {
                        GlStateManager.bindTexture(texItalicBold!!.glTextureId)
                        boldItalicChars
                    } else {
                        GlStateManager.bindTexture(texItalic!!.glTextureId)
                        italicChars
                    }
                } else if (colorIndex == 21) {
                    bold = false
                    italic = false
                    underline = false
                    strikethrough = false
                    GlStateManager.color(
                        (color shr 16 and 0xFF) / 255.0f, (color shr 8 and 0xFF) / 255.0f,
                        (color and 0xFF) / 255.0f, alpha
                    )
                    GlStateManager.bindTexture(tex.glTextureId)
                    currentData = charData
                }
                ++i
            } else if (character.code < currentData.size) {
                GL11.glBegin(4)
                drawChar(currentData, character, x.toFloat(), y.toFloat())
                GL11.glEnd()
                if (strikethrough) {
                    drawLine(
                        x,
                        y + currentData[character.code]!!.height / 2.0f,
                        x + currentData[character.code]!!.width - 8.0,
                        y + currentData[character.code]!!.height / 2.0f,
                        1.0f
                    )
                }
                if (underline) {
                    drawLine(
                        x,
                        y + currentData[character.code]!!.height - 2.0,
                        x + currentData[character.code]!!.width - 8.0,
                        y + currentData[character.code]!!.height - 2.0,
                        1.0f
                    )
                }
                x += (currentData[character.code]!!.width - 9 + charOffset).toDouble()
            }
            ++i
        }
        GL11.glPopMatrix()
        return x.toFloat() / 2.0f
    }

    override fun getStringWidth(text: String): Int {
        if (text == null) {
            return 0
        }
        var width = 0
        var currentData = charData
        var bold = false
        var italic = false
        val size = text.length
        var i = 0
        while (i < size) {
            val character = text[i]
            if (character == '\u00a7' && i < size) {
                val colorIndex = "0123456789abcdefklmnor".indexOf(character)
                if (colorIndex < 16) {
                    bold = false
                    italic = false
                } else if (colorIndex == 17) {
                    bold = true
                    currentData = if (italic) boldItalicChars else boldChars
                } else if (colorIndex == 20) {
                    italic = true
                    currentData = if (bold) boldItalicChars else italicChars
                } else if (colorIndex == 21) {
                    bold = false
                    italic = false
                    currentData = charData
                }
                ++i
            } else if (character.code < currentData.size && character >= '\u0000') {
                width += currentData[character.code]!!.width - 9 + charOffset
            }
            ++i
        }
        return width / 2
    }

    override fun setFont(font: Font) {
        super.setFont(font)
        setupBoldItalicIDs()
    }

    override fun setAntiAlias(antiAlias: Boolean) {
        super.setAntiAlias(antiAlias)
        setupBoldItalicIDs()
    }

    override fun setFractionalMetrics(fractionalMetrics: Boolean) {
        super.setFractionalMetrics(fractionalMetrics)
        setupBoldItalicIDs()
    }

    private fun setupBoldItalicIDs() {
        texBold = setupTexture(
            font.deriveFont(1), antiAlias, fractionalMetrics,
            boldChars
        )
        texItalic = setupTexture(
            font.deriveFont(2), antiAlias, fractionalMetrics,
            italicChars
        )
    }

    private fun drawLine(x: Double, y: Double, x1: Double, y1: Double, width: Float) {
        GL11.glDisable(3553)
        GL11.glLineWidth(width)
        GL11.glBegin(1)
        GL11.glVertex2d(x, y)
        GL11.glVertex2d(x1, y1)
        GL11.glEnd()
        GL11.glEnable(3553)
    }

    private fun setupMinecraftColorcodes() {
        var index = 0
        while (index < 32) {
            val noClue = (index shr 3 and 1) * 85
            var red = (index shr 2 and 1) * 170 + noClue
            var green = (index shr 1 and 1) * 170 + noClue
            var blue = (index shr 0 and 1) * 170 + noClue
            if (index == 6) {
                red += 85
            }
            if (index >= 16) {
                red /= 4
                green /= 4
                blue /= 4
            }
            colorCode[index] = red and 255 shl 16 or (green and 255 shl 8) or (blue and 255)
            ++index
        }
    }
}